### ISSUE TITLE
How to add support for swiglu in Megablocks?

### ISSUE THREAD

**Issue by fanshiqing** [stanford-futuredata/megablocks#35]
Hi @tgale96  @deepakn94,
During integration of Megablocks in to Megatron-LM, I found that when `--swiglu` ([src](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/arguments.py#L604-L605)) is enabled, the corresponding `ffn_hidden_size` is `int((4 * args.hidden_size * 2 / 3) / 64) * 64` instead of general `args.ffn_hidden_size = 4 * args.hidden_size` ([src](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/arguments.py#L243-L251)), this change will break the underlying assert in [dmoe.py](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/dmoe.py#L70):

`assert self.ffn_hidden_size % self.blocking == 0`

Could we add support for the case for swiglu where usually has **self.ffn_hidden_size % self.blocking !=0** ?  Thanks!

**Comment by tgale96** [comment]
Yes, we actually have someone looking into swiglu support this week, funnily enough.

Fwiw, you can get around that assertion by using the grouped MLP rather than the sparse MLP. You can do that by enabling [this](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/arguments.py#L41) flags.

**Comment by fanshiqing** [comment]
Thanks Trevor. 
One quick question: what's the difference of spraseMLP and GroupedMLP in Megablocks? What's their respective applicable scenarios? And what's their perf comparison for common MoE models? It seems these are not covered in the Megablocks paper.

**Comment by tgale96** [comment]
`SparseMLP` is what's described in the paper - it's still our recommended path, I would say. 

`GroupedMLP` is an alternative implementation based on a grouped GEMM kernel. We added it to work around the fact that Triton performance on Hopper is currently not very good, but the lack of SM90-optimzed grouped GEMM kernels in CUTLASS means that its performance is currently good mostly for the small expert per GPU regime (e.g., when you're training with expert model parallelism).

Using a grouped GEMM can be a little simpler - there are fewer constraints on the matrix dimension (e.g., 128-alignment with `SparseMLP`), no padding on the activation tensor and the metadata used to describe the sparsity is simpler (just token counts per expert). We're in touch with the CUTLASS team and once an SM90-optimized kernel is available (expected later this year) we should be able to get comparable performance to `SparseMLP` across the problem space and across different architectures.

I don't have an in-depth performance evaluation of the two yet (we're primarily using them for different scenarios/hardware platforms). Prior benchmarks of CUTLASS suggested we might lose a small amount of performance relative to the block-sparse approach but I need to study in more depth once the proper kernels are available and integrated.

One thing to note is that our `SparseMLP` can use less memory when `memory_optimized_mlp` is enabled. We're working on understanding why that is, since I expect both should be equivalent (and the same memory optimizations are supported for both). It seems like `GroupedMLP` interacts differently with PyTorch's caching allocator.

I Hope this helps! Let us know if there is anything we can do to help with your use case.

**Comment by tgale96** [comment]
Hello! FYI, we just merged swiglu support in for `SparseMLP` and `GroupedMLP`.

**Comment by bozheng-hit** [comment]
> Hello! FYI, we just merged swiglu support in for `SparseMLP` and `GroupedMLP`.

I cannot find the support for swiglu in this repository, could you please help point out how to enable swiglu?

**Comment by tgale96** [comment]
In your `Arguments` object, set `mlp_type="glu"` :)

**Comment by bozheng-hit** [comment]
> In your `Arguments` object, set `mlp_type="glu"` :)

That is actually not swiglu. But I have implemented myself. Thanks a lot anyway!

**Comment by tgale96** [comment]
Yes, it uses gelu instead of silu, I think. That's the difference you're concerned about?

**Comment by maxmatical** [comment]
@tgale96 I believe part of it is in addition to using silu, when using glu/swiglu, the `ffn_hidden_size` needs to be multiplied by `2` in the first mlp projection, like here: https://github.com/NVIDIA/Megatron-LM/blob/fab0bd693ec5be55b32c4f12a1ea44766ec63448/megatron/model/transformer.py#L92-L94

**Comment by maxmatical** [comment]
@tgale96 I think for swiglu [here](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/glu.py#L23) should be `args.ffn_hidden_size * 2`

**Comment by tgale96** [comment]
That'd be true if you fused the two matrices in the front half of the GLU. These implementations do not - you'll notice they have three weight matrices, the first two of which (w1/v1) effectively double `ffn_hidden_size`.

**Comment by maxmatical** [comment]
I see, then it should be a matter of replacing gelu with silu for swiglu

**Comment by tgale96** [comment]
Ya, we're going to make the activation function configurable for these ops. Fwiw, I don't expect silu versus gelu will make much of a difference given they're almost the same :)
